File: src/main/scala/com/github/mauricio/postgresql/parsers/DataRowParser.scala
package com.github.mauricio.postgresql.parsers

import org.jboss.netty.buffer.ChannelBuffer
import com.github.mauricio.postgresql.messages.backend.{DataRowMessage, Message}

/**
 * User: Maurício Linhares
 * Date: 3/4/12
 * Time: 10:56 AM
 */

object DataRowParser extends Decoder {

  def parseMessage(buffer: ChannelBuffer): Message = {

    val row = new Array[ChannelBuffer](buffer.readShort())

    0.until( row.length ).foreach {
      column =>
        val length = buffer.readInt()

        row(column) = if (length == -1) {
          null
        } else {
          val slice = buffer.slice( buffer.readerIndex(), length )
          buffer.readerIndex( buffer.readerIndex() + length )
          slice
        }
    }

    new DataRowMessage(row)
  }

}

File: postgresql-async/src/main/java/com/github/jasync/sql/db/postgresql/parsers/DataRowParser.kt
package com.github.jasync.sql.db.postgresql.parsers

import com.github.jasync.sql.db.postgresql.messages.backend.DataRowMessage
import com.github.jasync.sql.db.postgresql.messages.backend.ServerMessage
import io.netty.buffer.ByteBuf

object DataRowParser : MessageParser {

    override fun parseMessage(buffer: ByteBuf): ServerMessage {
        val row = Array(buffer.readShort().toInt()) {
            val length = buffer.readInt()
            if (length == -1) {
                null
            } else {
                buffer.readRetainedSlice(length)
            }
        }
        return DataRowMessage(row)
    }
}
